### ISSUE TITLE
CSRF is broken for certain actions like setting environment variables and restarting

### ISSUE THREAD

**Issue by zephyrweb** [kevana/ui-for-docker#208]
I get a 403 "Forbidden - CSRF token invalid" when this happens. Renaming containers works fine though, and that uses the CSRF header.

Do I need a specific version of docker for this?

**Comment by kevana** [comment]
That should work with any Docker version 1.6+. Those actions require multiple API calls, I'm guessing the CSRF token is getting stale somewhere in that process.

**Comment by zephyrweb** [comment]
Well, works fine on docker 1.11.1 on my mac. I have 1.7 on the server. I'll try upgrading.

**Comment by deviantony** [comment]
This problem just happened to me when I tried to remove a network. I switched to the image view, removed an image and then tried again to delete a network (and it worked this time).

**Comment by deviantony** [comment]
I reproduced the issue:

* Start UI-for-Docker
* Go to network view, try to delete network >  "Forbidden - CSRF token invalid
* Go to images view, delete an image > OK
* Go to network view, try to delete network > OK

**Comment by deviantony** [comment]
My issue was related to the container created with the `grunt run-dev` method. Sometimes, when adding changes to the app and reloading (thus creating a new container) a CSRF issue will happen.

I guess that is because the `authKey.dat` file is not persisted between these containers.

**Comment by kevana** [comment]
My intent when I added the authKey.dat was to persist it in a volume so CSRF wouldn't break every time it restarted but it looks like I never added the volume declaration to the Dockerfile or run commands.

**Comment by deviantony** [comment]
Yeah I'm currently looking at this one. But if we persist the `authKey.dat` file in a volume (e.g `/data` for example) it would run well in Docker but what about running it in standalone (e.g `./ui-for-docker` on a server or laptop)? It would require the existence of the folder on that machine.

**Comment by deviantony** [comment]
One solution is to add `-data` flag to specify the path to the data with a default value of `'.'`.

**Comment by deviantony** [comment]
I'm going to play with https://github.com/cloud-inovasi/cloudinovasi-ui/pull/22 to see if that works

**Comment by deviantony** [comment]
Do you recommend that the file `authKey.dat` should always be persisted and re-used?

**Comment by kevana** [comment]
I think so, not seeing any downside to keeping it around.

**Comment by deviantony** [comment]
Ok, persisted in a data volume should be fine.

Hey what do you think about opening a gitter room for this project? I still have a few questions that are not specially related to any issue here.

**Comment by kevana** [comment]
Done: https://gitter.im/kevana/ui-for-docker

**Comment by deviantony** [comment]
@zephyrweb any update on your side? Do you still have the issue on your server? As the authentication key file is not persisted at the moment, this issue may arises when you drop/recreate the ui-for-docker container.

It can also affects user using the `grunt` commands when developing locally.

**Comment by deviantony** [comment]
#225 should do the job. Feel free to re-open if needed.

**Comment by Tairy** [comment]
I get the same issue and I have done [#255](https://github.com/kevana/ui-for-docker/pull/225) but there is nothing changed!

**Comment by deviantony** [comment]
@Tairy I'll need more info. The only way to reproduce that issue is to access different instances of ui-for-docker using the same browser. So each instance will generate it's own CSRF data and the browser will be stuck on using the cookie of the first instance accessed.

If you want to access multiple instances without having any issues, you'll have to persist the CSRF data and share it amongst your ui-for-docker instances (use the `-v` flags or ask for more details).

You could also clean your cookies before trying to access a new instance, it will automatically retrieve the correct CSRF token after the first request.

**Comment by brunnels** [comment]
I'm having similar issues.  I'm unable to edit anything on the `#/containers/{container_id}/` pages.  When I click the commit button that would call `commit?container={container_id}&tag...`  I see `Forbidden - CSRF token invalid` in the XHR response in chrome.  Other pages work fine.  Running container like this

```docker run --name docker-ui -d -p 9000:9000 -v /tmp/ui-for-docker:/data -v /etc/docker/ssl/client:/certs uifd/ui-for-docker -H tcp://`hostname`.domain.com:2376 -tlsverify```

I don't know if it is related but I usually get no response in XHR from the log tail page at `#/containers/{container_id}/logs?stderr=0&stdout=1&tail=2000&timestamps=0`
